### ISSUE TITLE
Bookmarklet link on the 'Drummer FAQs' page

### ISSUE THREAD

**Issue by jgrisham** [scripting/drummerSupport#145]
The bookmarklet wouldn't initially work for me.

I discovered it on the following page, in the _[**Web clips & the bookmarklet**](http://docserver.scripting.com/drummer/general.opml#1634950078000)_ section:

> `http://docserver.scripting.com/drummer/general.opml`
> Drummer FAQs
> Last update: Monday November 29, 2021; 12:54 PM EST.

Perhaps there is some sort of _parsing?_ issue with one or more of the scripts at `docserver.scripting.com`. _(Alternatively, it could just be an issue with my computer/browser.)_

When a user _(at least in the `chromium`-based `Microsoft Edge` browser)_ tries to drag the bookmarklet to their 'bookmarks'/'favorites' bar, or right-clicks on the link and selects 'copy link', the following text is available:

_from [`general.opml`](http://docserver.scripting.com/drummer/general.opml)_
````JavaScript
javascript: (function () { function encode (s) { s = s.toString (); s = s.replace (new RegExp (
````

**Yup. It just stops.**

<details><summary>Details:</summary>

_(I added linebreaks in the codeblocks for readability when it wouldn't change the meaning)_

<details><summary>A) general.opml {xml / opml version}</summary>

<p />

_from [`general.opml?format=opml`](http://docserver.scripting.com/drummer/general.opml?format=opml)_

````xml
<outline 
    text="Install the bookmarklet by dragging the text below into your browser's chrome." 
    created="Sat, 23 Oct 2021 00:49:06 GMT">
    <outline 
        text="<a href=
            "javascript: (function () { function encode (s) { s = s.toString (); s = s.replace 
                (new RegExp ("%", "g"), " percent"); return (encodeURIComponent (s)); } 
                var url = "http://drummer.scripting.com/?"; 
                var what = "what=bookmarklet"; 
                var link = "&link=" + encode (location.href); 
                var title = "&title=" + encode (document.title); 
                var desc = "&description=" + encode (document.getSelection ()); 
                window.open (url + what + link + title + desc); }) (); "
            >
            <span style="font-family: Ubuntu; font-weight: bold; font-size: 20px;">
                Drummer
            </span></a>" 
        created="Sat, 23 Oct 2021 01:01:22 GMT"
    />
</outline>
````
   <tiny>_(highlighted as: `xml`)_</tiny>

----
</details>

<details><summary>B) general.opml {html version - view source}</summary>

<p />

 _(i.e. the HTML as delivered by the server; before any javascript is executed)_

<p />

_from [`general.opml`](http://docserver.scripting.com/drummer/general.opml)_

````JSON
{
    "text": "Install the bookmarklet by dragging the text below into your browser's chrome.",
    "created": "Sat, 23 Oct 2021 00:49:06 GMT",
    "subs": [
        {
            "text": "<a href=\"javascript: (function () { function encode (s) { s = s.toString (); s = s.replace (new RegExp (\"%\", \"g\"), \" percent\"); return (encodeURIComponent (s)); } var url = \"http://drummer.scripting.com/?\"; var what = \"what=bookmarklet\"; var link = \"&link=\" + encode (location.href); var title = \"&title=\" + encode (document.title); var desc = \"&description=\" + encode (document.getSelection ()); window.open (url + what + link + title + desc);  }) (); \"><span style=\"font-family: Ubuntu; font-weight: bold; font-size: 20px;\">Drummer</span></a>",
            "created": "Sat, 23 Oct 2021 01:01:22 GMT"
        }
    ]
},
````
   _(highlighted as: `JSON`)_

----
</details>

... at first glance nothing may seem to be amiss... 'double-quotes' appear to be 'escaped'

<details><summary>C) Everything above, but it appears in the _DOM_ as input to the browser's rendering engine</summary>

  _(i.e. after any javascripts have run; the HTML as rendered)_
<details><summary>(hidden for clarity: full version matching the above code blocks:)</summary>

general.opml {dom version - after any javascripts have run}
````HTML

<li>
    <div class="divOutlineText">
        <span class="spOutlineIcon"><a class="aOutlineWedgeLink" onclick="ecOutline (16)">
            <i class="fa fa-caret-down" style="color: silver;" id="idOutlineWedge16"></i>
        </a></span>
        <a class="aOutlineTextLink" onclick="ecOutline (16)">
            Install the bookmarklet by dragging the text below into your browser's chrome.
        </a>
        <a name="1634950146000"></a><span class="spNodePermalink"><a href="#1634950146000">#</a></span>
    </div>
    <ul class="ulOutlineList ulLevel3" id="idOutlineLevel16">
        <li><div class="divOutlineText">
            <a href="javascript: (function () { function encode (s) { s = s.toString (); s = s.replace (new RegExp (" %",="" "g"),="" "="" percent");="" return="" (encodeuricomponent="" (s));="" }="" var="" url="http://drummer.scripting.com/?" ;="" what="what=bookmarklet" link="&amp;link=" +="" encode="" (location.href);="" title="&amp;title=" (document.title);="" desc="&amp;description=" (document.getselection="" ());="" window.open="" (url="" desc);="" })="" ();=""><span style="font-family: Ubuntu; font-weight: bold; font-size: 20px;">Drummer</span></a>
            <a name="1634950882000"></a>
            <span class="spNodePermalink"><a href="#1634950882000">#</a></span>
        </div></li>
    </ul>
</li>

````
   _(highlighted as: `HTML`)_

<p />

general.opml {dom version - after any javascripts have run}
````JavaScript

<li>
    <div class="divOutlineText">
        <span class="spOutlineIcon"><a class="aOutlineWedgeLink" onclick="ecOutline (16)">
            <i class="fa fa-caret-down" style="color: silver;" id="idOutlineWedge16"></i>
        </a></span>
        <a class="aOutlineTextLink" onclick="ecOutline (16)">
            Install the bookmarklet by dragging the text below into your browser's chrome.
        </a>
        <a name="1634950146000"></a><span class="spNodePermalink"><a href="#1634950146000">#</a></span>
    </div>
    <ul class="ulOutlineList ulLevel3" id="idOutlineLevel16">
        <li><div class="divOutlineText">
            <a href="javascript: (function () { function encode (s) { s = s.toString (); s = s.replace (new RegExp (" %",="" "g"),="" "="" percent");="" return="" (encodeuricomponent="" (s));="" }="" var="" url="http://drummer.scripting.com/?" ;="" what="what=bookmarklet" link="&amp;link=" +="" encode="" (location.href);="" title="&amp;title=" (document.title);="" desc="&amp;description=" (document.getselection="" ());="" window.open="" (url="" desc);="" })="" ();=""><span style="font-family: Ubuntu; font-weight: bold; font-size: 20px;">Drummer</span></a>
            <a name="1634950882000"></a>
            <span class="spNodePermalink"><a href="#1634950882000">#</a></span>
        </div></li>
    </ul>
</li>

`````
   _(highlighted as: `JavaScript`)_

----
</details>

1. The single line in question; highlighted as `javascript`:
````JavaScript
<a href="javascript: (
    function () { 
        function encode (s) { 
            s = s.toString (); 
            s = s.replace (new RegExp (" %",="" "g"),="" "="" percent");="" 
            return="" (encodeuricomponent="" (s));="" }="" 
        var="" url="http://drummer.scripting.com/?" ;="" 
        what="what=bookmarklet" 
        link="&amp;link=" +="" encode="" (location.href);="" 
        title="&amp;title=" (document.title);="" 
        desc="&amp;description=" (document.getselection="" ());="" 
        window.open="" (url="" desc);="" })="" ();="">
    <span style="font-family: Ubuntu; font-weight: bold; font-size: 20px;">Drummer</span>
</a>
````

2. The single line in question; highlighted as `HTML`:
````HTML
<a href="javascript: (
    function () { 
        function encode (s) { 
            s = s.toString (); 
            s = s.replace (new RegExp (" %",="" "g"),="" "="" percent");="" 
            return="" (encodeuricomponent="" (s));="" }="" 
        var="" url="http://drummer.scripting.com/?" ;="" 
        what="what=bookmarklet" 
        link="&amp;link=" +="" encode="" (location.href);="" 
        title="&amp;title=" (document.title);="" 
        desc="&amp;description=" (document.getselection="" ());="" 
        window.open="" (url="" desc);="" })="" ();="">
    <span style="font-family: Ubuntu; font-weight: bold; font-size: 20px;">Drummer</span>
</a>
````
----
</details>

AFAIK, the working bookmarklet code is:
````JavaScript
javascript: (function () { function encode (s) { s = s.toString (); s = s.replace (new RegExp ("%", "g"), " percent"); return (encodeURIComponent (s)); } var url = "http://drummer.scripting.com/?"; var what = "what=bookmarklet"; var link = "&link=" + encode (location.href); var title = "&title=" + encode (document.title); var desc = "&description=" + encode (document.getSelection ()); window.open (url + what + link + title + desc); }) ();
````

<details><summary>Same thing; nicer formatting:</summary>

````JavaScript
javascript: (
    function () { 
        function encode (s) { 
            s = s.toString (); 
            s = s.replace (new RegExp ("%", "g"), " percent"); 
            return (encodeURIComponent (s)); 
            } 
        var url = "http://drummer.scripting.com/?"; 
        var what = "what=bookmarklet"; 
        var link = "&link=" + encode (location.href); 
        var title = "&title=" + encode (document.title); 
        var desc = "&description=" + encode (document.getSelection ()); 
        window.open (url + what + link + title + desc); 
        }
    ) 
    ();
````

</details>

</details>

----

## Tl;dr:
I was able to make the bookmarklet work by directly copying the code from the `opml`-formatted [page](http://docserver.scripting.com/drummer/general.opml?format=opml), but I'm still posting this so that:
 1. the FAQ page can be updated so that the bookmarklet installation works as described:
    > "Install the bookmarklet by dragging the text below into your browser's chrome."
 2. in the event that there's a broader parsing issue with how quote marks _(or other special characters)_ that exist in the OPML source are rendered in the browser. 
   a. _(perhaps this ends up being a `docserver` issue more than an actual `Drummer` issue?)_

_P.S. Thanks for all the work you are doing and for sharing your ideas / creations with the world!_

**Comment by jgrisham** [comment]
Tangentially...

... while you're looking at that page anyway, the link to `DocServer` in the **["Your own DocServer"](http://docserver.scripting.com/drummer/general.opml#1632846536000)** section uses `https` instead of `http` in its URL.

**Comment by scripting** [comment]
@jgrisham -- thanks for the very detailed post. I had a lot of help getting the bookmarklet to work, so what I'm going to do is post a link to this issue in that thread, so the experts can read this and tell me what to do. :smile:

And thanks for the heads-up on the link to the DocServer app. I fixed it.

**Comment by jgrisham** [comment]
Of course! _(Apologies for not making it more concise, but it was getting late.)_



----
For the sake of anyone _'newly entering the chat'_:

- To be clear, the _bookmarklet code itself_ appears to work fine.
<details><summary>Example:</summary>

The following page, when loaded in to a browser, produces a working bookmarklet _and_ passes the [W3C Markup Validation Service parser](https://validator.w3.org/):

````html
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en">
  <head>
    <meta charset="utf-8" />
    <title>Drummer bookmarklet demo</title>
    <link rel="stylesheet" href="style.css">
    <script src="script.js"></script>
  </head>
  <body>
    <a href="javascript: (function () { function encode (s) { s = s.toString (); s = s.replace (new RegExp (&quot;%&quot;, &quot;g&quot;), &quot; percent&quot;); return (encodeURIComponent (s)); } var url = &quot;http://drummer.scripting.com/?&quot;; var what = &quot;what=bookmarklet&quot;; var link = &quot;&link=&quot; + encode (location.href); var title = &quot;&title=&quot; + encode (document.title); var desc = &quot;&description=&quot; + encode (document.getSelection ()); window.open (url + what + link + title + desc); }) ();">Send to Drummer</a>
  </body>
</html>
````
</details>

What appears to be amiss is the whatever client-side code that converts strings in from the in-line JSON _(containing quote marks)_ for display in HTML.

Unfortunately, there are multiple levels of parsing happening here, across several different formats, each potentially with their own set of special requirements:
 1. OPML / XML _(server-side, in this case?)_
 2. JSON
 3. HTML
 4. JavaScript

It appears that the OPML -> JSON conversion is correctly quoting the characters _(presumably server-side in `DocServer` / `PagePark`)_.

If I had to guess _(and to be clear, I know nothing about `JavaScript`)_, the client-side is un-quoting the strings when parsing the JSON but neglecting to re-quote it when injecting new HTML into the page's DOM. _(e.g. some of the `var` tokens are just missing, implying that the JS engine actually tried to execute those portions of the string)_

... but... since OPML _is_ XML, [shouldn't it be escaping double-quotes](https://www.onwebsecurity.com/security/properly-encoding-and-escaping-for-the-web.html) (`"`) as either `&quot;` or `&#34;` for [any text stored in value attributes](https://stackoverflow.com/questions/1091945/what-characters-do-i-need-to-escape-in-xml-documents) that are themselves delimited by double-quotes, well before being turned into in-line `JSON`?

<p /><br />
Great. So, I've narrowed it down to either _client-side_ or _server-side_. :( 

\*steps back slowly\*

----

Good luck to whomever figures out a fix for this, and thank you!

**Comment by jgrisham** [comment]
Until the actual root cause of this is discovered and corrected _(since I'm assuming that incorrectly formatted OPML should never actually be allowed to cause unintended JavaScript execution in the browser, not to mention possible user data loss)_, <details><summary>…a potential workaround:</summary>

- simply replacing each instance of the double-quotes `"` in the JavaScript of the bookmarklet with single-quotes `'` appears to validate _both_ as [proper](https://beautifytools.com/javascript-validator.php) `JavaScript` and as [valid contents](https://stackoverflow.com/a/1091953/5711986) of a `"`-delimited `value` of an XML tag:

````Javascript
javascript: (function () { function encode (s) { s = s.toString (); s = s.replace (new RegExp ('%', 'g'), ' percent'); return (encodeURIComponent (s)); } var url = 'http://drummer.scripting.com/?'; var what = 'what=bookmarklet'; var link = '&link=' + encode (location.href); var title = '&title=' + encode (document.title); var desc = '&description=' + encode (document.getSelection ()); window.open (url + what + link + title + desc); }) ();
````

e.g., replace the outline text so that the `OPML` matches the following:

````xml
<outline 
    text="Install the bookmarklet by dragging the text below into your browser's chrome." 
    created="Sat, 23 Oct 2021 00:49:06 GMT">
	<outline 
     	   text="<a href="javascript: (function () { function encode (s) { s = s.toString (); s = s.replace (new RegExp ('%', 'g'), ' percent'); return (encodeURIComponent (s)); } var url = 'http://drummer.scripting.com/?'; var what = 'what=bookmarklet'; var link = '&link=' + encode (location.href); var title = '&title=' + encode (document.title); var desc = '&description=' + encode (document.getSelection ()); window.open (url + what + link + title + desc); }) ();">
		<span style="font-family: Ubuntu; font-weight: bold; font-size: 20px;">
        	        Drummer
  		</span></a>" 
     	   created="Sat, 23 Oct 2021 01:01:22 GMT"
  	  />
</outline>
````

... but it should _probably_ still be something like this, right?:

````xml
<outline 
    text="Install the bookmarklet by dragging the text below into your browser's chrome." 
    created="Sat, 23 Oct 2021 00:49:06 GMT">
	<outline 
     	   text="<a href=&quot;javascript: (function () { function encode (s) { s = s.toString (); s = s.replace (new RegExp ('%', 'g'), ' percent'); return (encodeURIComponent (s)); } var url = 'http://drummer.scripting.com/?'; var what = 'what=bookmarklet'; var link = '&link=' + encode (location.href); var title = '&title=' + encode (document.title); var desc = '&description=' + encode (document.getSelection ()); window.open (url + what + link + title + desc); }) ();&quot;>
		<span style=&quot;font-family: Ubuntu; font-weight: bold; font-size: 20px;&quot;>
        	        Drummer
  		</span></a>" 
     	   created="Sat, 23 Oct 2021 01:01:22 GMT"
  	  />
</outline>
````
----
_On the bright side, at least we're nowhere near `EBCDIC`-encoded XML, piped to a `bash` script by way of some `php`? 😭_

</details>

**Comment by Kelly-NAproducR** [comment]
Dave, I've gone back to exploring Drummer.  I've never been able to get the Drummer bookmarklet to work.  I look at the code, and it appears incomplete to me, base on what I have used in the past.  Is this complete, working code?

javascript: (function () { function encode (s) { s = s.toString (); s = s.replace (new RegExp (

I use Brave Browser and I have tried Chrome, Firefox, Vivaldi and Opera.  All failed.

Thank you!

**Comment by scripting** [comment]
@Kelly-NAproducR -- hopefully someone else can help with this. I am not able to get into this now. But anyone could make a bookmarklet for Drummer, it doesn't have to be me.

**Comment by jgrisham** [comment]
@Kelly-NAproducR The code  you listed is _not_ correct _(or more precisely, it is not complete)_.

<details>
<summary>Background:</summary>
<br />

Considering that most structured languages require that grouping symbols _(such as `()` / `[]` / `{}` / `''` / `""`)_ be used in matched pairs _(much as HTML requires tags to be 'closed')_, ...

... the fact that the javascript snippet posted ends in a lonely `(` is a pretty big indicator that the text has been clipped.

<hr />

</details>

I haven't tested these recently[1], but here are the complete versions of the bookmarklet code <s>_(copied verbatim)_</s>[2] from the various posts earlier in this conversation:

<details><summary>... from post no. 1</summary>

([2021-12-21 - Post #1](https://github.com/scripting/drummerSupport/issues/145#issue-1085452737))

> AFAIK, the working bookmarklet code is:
> ````JavaScript
> javascript: (function () { function encode (s) { s = s.toString (); s = s.replace (new RegExp ("%", "g"), " percent"); return (encodeURIComponent (s)); } var url = "https://drummer.land/?"; var what = "what=bookmarklet"; var link = "&link=" + encode (location.href); var title = "&title=" + encode (document.title); var desc = "&description=" + encode (document.getSelection ()); window.open (url + what + link + title + desc); }) ();
> ````
> 
> <details><summary>Same thing; nicer formatting:</summary>
> 
> ````JavaScript
> javascript: (
>     function () { 
>         function encode (s) { 
>             s = s.toString (); 
>             s = s.replace (new RegExp ("%", "g"), " percent"); 
>             return (encodeURIComponent (s)); 
>             } 
>         var url = "https://drummer.land/?"; 
>         var what = "what=bookmarklet"; 
>         var link = "&link=" + encode (location.href); 
>         var title = "&title=" + encode (document.title); 
>         var desc = "&description=" + encode (document.getSelection ()); 
>         window.open (url + what + link + title + desc); 
>         }
>     ) 
>     ();
> ````
> 
> </details>

</details>


<details><summary>... from post no. 4</summary>

([2021-12-21 - Post #4](https://github.com/scripting/drummerSupport/issues/145#issuecomment-999154237))

> 
> The following page, when loaded in to a browser, produces a working bookmarklet _and_ passes the [W3C Markup Validation Service parser](https://validator.w3.org/):
> 
> ````html
> <!DOCTYPE html>
> <html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en">
>   <head>
>     <meta charset="utf-8" />
>     <title>Drummer bookmarklet demo</title>
>     <link rel="stylesheet" href="style.css">
>     <script src="script.js"></script>
>   </head>
>   <body>
>     <a href="javascript: (function () { function encode (s) { s = s.toString (); s = s.replace (new RegExp (&quot;%&quot;, &quot;g&quot;), &quot; percent&quot;); return (encodeURIComponent (s)); } var url = &quot;https://drummer.land/?&quot;; var what = &quot;what=bookmarklet&quot;; var link = &quot;&link=&quot; + encode (location.href); var title = &quot;&title=&quot; + encode (document.title); var desc = &quot;&description=&quot; + encode (document.getSelection ()); window.open (url + what + link + title + desc); }) ();">Send to Drummer</a>
>   </body>
> </html>
> ````
> </details>

</details>



<details><summary>... from post no. 5</summary>

([2021-12-21 - Post #5](https://github.com/scripting/drummerSupport/issues/145#issuecomment-999201377))

> ````Javascript
> javascript: (function () { function encode (s) { s = s.toString (); s = s.replace (new RegExp ('%', 'g'), ' percent'); return (encodeURIComponent (s)); } var url = 'https://drummer.land/?'; var what = 'what=bookmarklet'; var link = '&link=' + encode (location.href); var title = '&title=' + encode (document.title); var desc = '&description=' + encode (document.getSelection ()); window.open (url + what + link + title + desc); }) ();
> ````

</details>


<s>_(I'd try one at a time, starting from the last one, until you find one that works. See the linked posts above for more context about each.)_</s>

<hr />

### Tl;dr:

Try this one _(it's the same as number 5 above)_; it works for me on Firefox _(v120 running on Fedora Linux)_:


> ````Javascript
> javascript: (function () { function encode (s) { s = s.toString (); s = s.replace (new RegExp ('%', 'g'), ' percent'); return (encodeURIComponent (s)); } var url = 'https://drummer.land/?'; var what = 'what=bookmarklet'; var link = '&link=' + encode (location.href); var title = '&title=' + encode (document.title); var desc = '&description=' + encode (document.getSelection ()); window.open (url + what + link + title + desc); }) ();
> ````

<br />

Bonus: if you have text selected on the page when you activate the bookmarklet, it will be captured as the 'description' and appear as a child of the bookmark in the Drummer outline.

 _(This is probably documented, but I had forgotten about this useful feature until just now!)_



<hr />

[Footnote 1]: _(those posts are over 26 months old, so I can't guarantee that Drummer hasn't changed something important in the meantime)_ 
* _(Edit: something did change -- the URL used for Drummer!)_

[Footnote 2]: Thanks to @troutcolor, they are now direct copies _except_ that I changed every instance of <var>http://drummer.scripting.com</var> to <var>https://drummer.land</var>. _(reply edited 2024-03-14)_

**Comment by troutcolor** [comment]
Hi,
I checked my bookmarklet which seemed to be broken, but it was pointing at the old url. I change it to this one:

`javascript: (function () { function encode (s) { s = s.toString (); s = s.replace (new RegExp ("%", "g"), " percent"); return (encodeURIComponent (s)); } var url = "https://drummer.land/?"; var what = "what=bookmarklet"; var link = "&link="   encode (location.href); var title = "&title="   encode (document.title); var desc = "&description="   encode (document.getSelection ()); window.open (url   what   link   title   desc);  }) ();`

Which worked as expected. 
I don't recall where I got it or if I'd edited it. It was url encoded when I copied it I unencode it hope it works for you

cheers
john
